src/rust/catmem/queue.rs

```diff
@@ -5,68 +5,292 @@
+// Constants
+//======================================================================================================================
+
+/// Maximum number of retries for pushing a EoF signal.
+const MAX_RETRIES_PUSH_EOF: u32 = 16;
 
 //======================================================================================================================
 // Structures
 //======================================================================================================================
 
-/// Per-queue metadata: reference to the shared pipe.
+/// This structure contains code for manipulating a single, Catmem-specific Demikernel queue. Queue state is kept in
+/// the [ring] structure, while [pending_ops] holds the map of TaskHandles and YielderHandle for currently active async
+/// functions.
+#[derive(Clone)]
 pub struct CatmemQueue {
-    pipe: Pipe,
-    pending_ops: HashMap<TaskHandle, YielderHandle>,
+    ring: Rc<RefCell<Ring>>,
+    pending_ops: Rc<RefCell<HashMap<TaskHandle, YielderHandle>>>,
 }
 
 //======================================================================================================================
 // Associated Functions
 //======================================================================================================================
 
 impl CatmemQueue {
-    pub fn new(ring: SharedRingBuffer<u16>) -> Self {
-        Self {
-            pipe: Pipe::new(ring),
-            pending_ops: HashMap::<TaskHandle, YielderHandle>::new(),
+    /// This function creates a new CatmemQueue and a new shared ring buffer and connects to it to either the consumer
+    /// or producer end indicated by [mode].
+    pub fn create(name: &str, mode: QMode) -> Result<Self, Fail> {
+        match mode {
+            QMode::Push => Ok(Self {
+                ring: Rc::new(RefCell::<Ring>::new(Ring::create_push_ring(name)?)),
+                pending_ops: Rc::new(RefCell::<HashMap<TaskHandle, YielderHandle>>::new(HashMap::<
+                    TaskHandle,
+                    YielderHandle,
+                >::new(
+                ))),
+            }),
+            QMode::Pop => Ok(Self {
+                ring: Rc::new(RefCell::<Ring>::new(Ring::create_pop_ring(name)?)),
+                pending_ops: Rc::new(RefCell::<HashMap<TaskHandle, YielderHandle>>::new(HashMap::<
+                    TaskHandle,
+                    YielderHandle,
+                >::new(
+                ))),
+            }),
+        }
+    }
+
+    /// This function creates a new CatmemQueue and attaches to an existing share ring buffer as either a consumer or
+    /// producer as indicated by [mode].
+    pub fn open(name: &str, mode: QMode) -> Result<Self, Fail> {
+        match mode {
+            QMode::Push => Ok(Self {
+                ring: Rc::new(RefCell::<Ring>::new(Ring::open_push_ring(name)?)),
+                pending_ops: Rc::new(RefCell::<HashMap<TaskHandle, YielderHandle>>::new(HashMap::<
```

Comment:
The creation of pending_ops is repeated. Possibly, it can be moved outside of the match statements?

Reply:
Sure.